enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,441 @@
+and route will serve the default certificates that are generated by cluster-ingress-operator.
+If the `.spec.tls.certificate` and `.spec.tls.key` are replaced as well, then the
+controller will update the `.status.certificate` to `CustomCertificate`
+
+The new secret-injector controller will not reconcile Routes that are owned by the
+ingress-to-route controller.
+
+### Risks and Mitigations
+
+N.A
+
+### Drawbacks
+
+This introduces an inconsistency between Ingress and Route CRD with respect to how
+a secret reference can be provided with the certificate data. The ingress has a field
+`.spec.tls.secretName` where as the Route will have an annotation.
+
+## Design Details
+
+### Open Questions [optional]
+
+1. Does this warrant making changes to the ingress-to-route controller in how
+   `.spec.tls.secretName` is processed? Currently, the controller reads the secret
+   data and copies it over to the Route that is created. Since this enhancement
+   introduces a difference between managed certificate provided via secrets and  
+   manually provided secrets, this distinction isn't possible with Routes created
+   via Ingresses. [CLOSED]
+
+   - Proposed change
+     The ingress-to-route controller will also process the secret-reference annotation and if
+     present copies it over to the route that is created. This annotation will take
+     precedence over `.spec.tls.secretName` and this offers enough distinction between
+     third-party/manual certificate management.
+
+**Answer**: The behaviour of the ingress-to-route controller will be retained as is and
+no changes will be done.
+
+2. How/where to implement e2e tests for [route-controller-manager](https://github.com/openshift/route-controller-manager)?
+   [CLOSED]
+
+**Answer**: The e2e tests will be added to openshift/origin.
+
+### Test Plan
+
+This enhancement will be tested in isolation of cert-manager-operator as part of
+core OCP payload. This will also be tested with cert-manager-operator as part of the
+operators e2e test suite.
+
+1. Test Route API without interfacing with Ingresses
+
+   a. Create a edge terminated route that is using default certificates.
+   b. Update route with the secret-reference annotation and ensure the `.spec.tls` and `.status`
+   is updated.
+   c. Verify that the associated openshift routers are also serving the updated
+   certificates and not the default certificates.
+   d. Update certificate data and verify if the certificates served by the router
+   are updated as well.
+   d. Delete the annotation and ensure that the route `.spec.tls` and `.status`
+   denotes the usage of default certificates. Also verify the associated
+   openshift router is serving the default certificates.
+
+Various other transitions, `Custom`->`Managed`-`Custom` will also be tested.
+
+### Graduation Criteria
+
+This is a user facing change and will directly go to GA. This feature requires an
```

Comment:
Why not techpreview? Is there no merit in developing this in a techpreview environment and enabling it later? Individual API fields can be added as TechPreview with our current tooling